primitives/src/merkle_tree/mod.rs

```diff
@@ -0,0 +1,288 @@
+            )),
+            LookupResult::EmptyLeaf(_) => Err(PrimitivesError::InternalError(
+                "Expected Ok, found EmptyLeaf".to_string(),
+            )),
+        }
+    }
+
+    /// Assert the lookup result is NotInMemory. Return a non-membership proof.
+    pub fn expect_empty(self) -> Result<N, PrimitivesError> {
+        match self {
+            LookupResult::EmptyLeaf(n) => Ok(n),
+            LookupResult::Ok(..) => Err(PrimitivesError::InternalError(
+                "Expected EmptyLeaf, found element".to_string(),
+            )),
+            LookupResult::NotInMemory => Err(PrimitivesError::InternalError(
+                "Expected EmptyLeaf, found NotInMemory".to_string(),
+            )),
+        }
+    }
+}
+
+/// An element of a Merkle tree.
+pub trait Element: Clone + Eq + PartialEq {}
+impl<T: Clone + Eq + PartialEq> Element for T {}
+
+/// An index type of a leaf in a Merkle tree.
+pub trait Index: Debug + Eq + PartialEq + Ord + PartialOrd + Clone + ToTraversalPath {}
+impl<T: Debug + Eq + PartialEq + Ord + PartialOrd + Clone + ToTraversalPath> Index for T {}
+
+/// An internal node value type in a Merkle tree.
+pub trait NodeValue: Default + Eq + PartialEq + Copy + Clone + Debug {}
+impl<T: Default + Eq + PartialEq + Copy + Clone + Debug> NodeValue for T {}
+
+/// Merkle tree hash function
+pub trait DigestAlgorithm<E, I, T>
+where
+    E: Element,
+    I: Index,
+    T: NodeValue,
+{
+    // Possible improvement: adding digest_element() and digest_index()
+    /// Digest a list of values
+    fn digest(data: &[T]) -> T;
+
+    /// Digest a leaf (an indexed element)
+    fn digest_leaf(pos: &I, elem: &E) -> T;
+}
+
+/// An trait for Merkle tree index type.
+pub trait ToTraversalPath {
+    /// Convert the given index to a vector of branch indices given tree height
+    /// and arity.
+    fn to_traverse_path(&self, height: usize, arity: usize) -> Vec<usize>;
+}
+
+impl<F: Into<BigUint> + Clone> ToTraversalPath for F {
+    fn to_traverse_path(&self, height: usize, arity: usize) -> Vec<usize> {
+        let mut pos: BigUint = <F as Into<BigUint>>::into(self.clone());
+        let mut ret = vec![];
+        for _i in 0..height {
+            ret.push((&pos % (arity as u64)).to_usize().unwrap());
+            pos /= arity as u64;
+        }
+        ret
+    }
+}
+
+/// A merkle commitment consists a root hash value, a tree height and number of
+/// leaves
+#[derive(Eq, PartialEq, Clone, Copy, Serialize, Deserialize)]
+pub struct MerkleCommitment<T: NodeValue> {
+    /// Root of a tree
+    pub root_value: T,
+    /// Height of a tree
+    pub height: usize,
+    /// Number of leaves in the tree
+    pub num_leaves: u64,
+}
+
+/// Basic functionalities for a merkle tree implementation. Abstracted as an
+/// accumulator for fixed-length array. Supports generate membership proof at a
+/// given position and verify a membership proof.
+pub trait MerkleTreeScheme: Sized {
```

Comment:
I wonder if we should call this (and the whole module, maybe) something like `AccumulatorScheme`, to be a bit more generic

Reply:
Good point. If we do this, then everything would be in a different flavor. Probably we should also change `AppendableMerkleTreeScheme` to `VectorAccumulatorScheme` and `UniversalMerkleTreeScheme` to `MapAccumulatorScheme`? 

cc @alxiong @chancharles92 @tessico @fkrell @tessico

Reply:
1. If we want to make it an `AccumulatorScheme` trait, then we should remove `height()` API, remove ARITY constant, and rename some APIs as well (e.g. root(), num_leaves(), etc). Also, we should move the `AccumulatorScheme` trait to another file I think.
2. I still prefer to keep `MerkleTreeScheme`, `AppendableMerkleTreeScheme`, and `UniversalMerkleTreeScheme` traits, as their possible instantiations are rather different from other flavors of instantiations (e.g. bilinear-map or RSA accumulators). If we want `AccumulatorScheme` trait, what we can do is add `trait MerkleTreeScheme : AccumulatorScheme` after defining the `AccumulatorScheme` trait.